lld/MachO/ConcatOutputSection.cpp

```diff
@@ -116,6 +116,30 @@ void ConcatOutputSection::addInput(ConcatInputSection *input) {
 
 DenseMap<Symbol *, ThunkInfo> lld::macho::thunkMap;
 
+// Returns the target Symbol that a relocation refers to.
+// A Reloc can refer to either a Symbol directly, or to an InputSection.
+// For InputSection referents, we return the first Symbol at offset 0.
+// This conversion is necessary because the thunk generation algorithm
+// can only handle Symbols as branch targets, not InputSections.
+static Symbol *getReferentSymbol(const Reloc &r) {
+  if (auto *sym = r.referent.dyn_cast<Symbol *>()) {
```

Comment:
We are migrating away from `PointerUnion::dyn_cast`:

https://github.com/llvm/llvm-project/blob/c89735d289f341985ca2ea74486b96bc611b3c64/llvm/include/llvm/ADT/PointerUnion.h#L146-L147

AFAICT from `llvm_unrechable` below, you expect `r.referent.isNull()` to be false, so may I suggest the following?

```suggestion
  if (auto *sym = llvm::dyn_cast<Symbol *>(r.referent)) {
```